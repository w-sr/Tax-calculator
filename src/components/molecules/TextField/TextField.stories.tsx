import type { Meta, StoryObj } from "@storybook/react";

import { TextField } from "./TextField";

const meta: Meta<typeof TextField> = {
  title: "TextField",
  component: TextField,
};

export default meta;

type Story = StoryObj<typeof TextField>;

export const WithLabel: Story = {
  render: () => <TextField label="Text" />,
};

export const WithoutLabel: Story = {
  render: () => <TextField />,
};

export const WithError: Story = {
  render: () => <TextField label="Text" helper="Text is required" />,
};
