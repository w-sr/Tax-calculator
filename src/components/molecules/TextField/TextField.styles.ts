import styled, { css } from "styled-components";

export const TextField = styled.div`
  display: flex;
  align-items: center;
`;

export const InputWrap = styled.div`
  position: relative;
  flex: 1;
`;

export const Label = styled.label`
  ${({ theme }) => css`
    width: 100px;
    font-size: ${theme.font.small.size};
    line-height: ${theme.font.small.lineHeight};
    font-weight: ${theme.font.small.weight};
    text-transform: capitalize;
  `}
`;

export const Input = styled.input`
  ${({ theme }) => css`
    width: 100%;
    padding: 8px;
    font-size: ${theme.font.small.size};
    line-height: ${theme.font.small.lineHeight};
    font-weight: ${theme.font.small.weight};
    outline: none;
    box-sizing: border-box;
  `}
`;

export const HelperText = styled.p`
  ${({ theme }) => css`
    position: absolute;
    margin: 0;
    margin: 4px 0;
    font-size: ${theme.font.tiny.size};
    line-height: ${theme.font.tiny.lineHeight};
    font-weight: ${theme.font.tiny.weight};
    color: ${theme.colors.red};
  `}
`;
