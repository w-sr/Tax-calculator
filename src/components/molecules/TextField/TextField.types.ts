import type { InputHTMLAttributes } from "react";

export type TextFieldRef = HTMLInputElement;

export type TextFieldProps = InputHTMLAttributes<HTMLInputElement> & {
  helper?: string;
  label?: string;
  onChange?: (e: React.ChangeEvent<HTMLElement>) => void;
};
