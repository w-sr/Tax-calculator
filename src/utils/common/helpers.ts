const buildPercent = (decimal?: number) =>
  new Intl.NumberFormat("en-US", {
    style: "percent",
    minimumFractionDigits: decimal ?? 2,
  });

/**
 *
 * @param value
 * @param decimal
 * @returns value with percentage
 * @example 0.20 => 20%
 */
export const percentValue = (
  value: number | undefined,
  decimal?: number
): string => {
  if (value === undefined) return "N/A";
  const percent = buildPercent(decimal);
  return percent.format(value);
};

/**
 *
 * @param value
 * @param decimal
 * @returns value with currency
 * @example 20 => $20
 */
export const moneyValue = (
  value: number | undefined,
  decimal?: number
): string => {
  const money = new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
    currencyDisplay: "narrowSymbol",
    minimumFractionDigits: decimal ?? 2,
  });
  return money.format(value ?? 0);
};

/**
 *
 * @param min
 * @param max
 * @returns formatted bracket
 * @example min: 0, max: 2000 => "$2000 or less"
 *          min: 2000, max: 4000 => "$2000 to $4000"
 *          min: 4000 => "More than $4000"
 */
export const bracketFormat = (min: number, max?: number): string => {
  if (min === 0) return moneyValue(max, 0) + " or less";
  if (max) return moneyValue(min, 0) + " to " + moneyValue(max, 0);
  else return "More than " + moneyValue(min, 0);
};
