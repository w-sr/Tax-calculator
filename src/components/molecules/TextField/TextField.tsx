import { forwardRef } from "react";
import { TextFieldProps, TextFieldRef } from "./TextField.types";
import * as S from "./TextField.styles";

const TextField = forwardRef<TextFieldRef, TextFieldProps>(
  ({ label, helper, ...props }, ref) => {
    return (
      <S.TextField>
        <S.Label>{label}</S.Label>
        <S.InputWrap>
          <S.Input ref={ref} {...props} />
          {helper && <S.HelperText>{helper}</S.HelperText>}
        </S.InputWrap>
      </S.TextField>
    );
  }
);

TextField.displayName = "TextField";

export { TextField };
