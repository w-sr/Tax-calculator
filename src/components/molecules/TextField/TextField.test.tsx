import { renderWithTheme } from "../../../utils/tests/helpers";
import { TextField } from "./TextField";

describe("<TextField />", () => {
  it("should match the snapshot", () => {
    const { container } = renderWithTheme(<TextField />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
