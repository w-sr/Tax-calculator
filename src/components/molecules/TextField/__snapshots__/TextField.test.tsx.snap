// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should match the snapshot 1`] = `
<div
  class="sc-beqWaB iHXrTl"
>
  <label
    class="sc-dmqHEX floUjU"
  />
  <div
    class="sc-gueYoa jjURQE"
  >
    <input
      class="sc-hLseeU jCKmqV"
    />
  </div>
</div>
`;
